### PR TITLE
Raise DatasetTooSmall exception if canonical nodes is less than num samples

### PR DESCRIPTION
**Pull request by irenedea**
Raises a specialized error if you init a dataset that has too few samples.

### CODE REVIEW

File: llmfoundry/command_utils/data_prep/convert_text_to_mds.py [hunk 1]
```diff
@@ -478,7 +478,9 @@ def convert_text_to_mds(
     index_path = os.path.join(local_output_folder, 'index.json')
     with open(index_path, 'r') as index_file:
         if not json.load(index_file)['shards']:
-            raise DatasetTooSmallError()
+            raise DatasetTooSmallError(
+                reason='No shards were created when convering text to mds.',
```
Comment (snarayan21) [inline llmfoundry/command_utils/data_prep/convert_text_to_mds.py:482]:
```suggestion
                reason='No shards were created when convering text to MDS.',
```

File: llmfoundry/command_utils/data_prep/convert_text_to_mds.py [hunk 2]
```diff
@@ -478,7 +478,9 @@ def convert_text_to_mds(
     index_path = os.path.join(local_output_folder, 'index.json')
     with open(index_path, 'r') as index_file:
         if not json.load(index_file)['shards']:
-            raise DatasetTooSmallError()
+            raise DatasetTooSmallError(
```
Comment (snarayan21) [inline llmfoundry/command_utils/data_prep/convert_text_to_mds.py:481]: this exception already existed before right? just making sure, for my understanding

File: llmfoundry/data/finetuning/tasks.py [hunk 3]
```diff
@@ -1033,7 +1034,16 @@ def build_from_streaming(
         *args: Any,
         **kwargs: Any,
     ) -> StreamingFinetuningDataset:
-        return self.streaming_dataset_class(*args, **kwargs)
+        dataset = self.streaming_dataset_class(*args, **kwargs)
+        num_canonical_nodes = dataset.num_canonical_nodes
+        num_samples = dataset.num_samples
+        if num_canonical_nodes is not None and num_canonical_nodes < num_samples:
+            raise DatasetTooSmallError(
```
Comment (snarayan21) [inline llmfoundry/data/finetuning/tasks.py:1041]: Is this something that's actionable by the user? isn't the index.json not accessible by them?

File: llmfoundry/data/finetuning/tasks.py [hunk 4]
```diff
@@ -1033,7 +1034,16 @@ def build_from_streaming(
         *args: Any,
         **kwargs: Any,
     ) -> StreamingFinetuningDataset:
-        return self.streaming_dataset_class(*args, **kwargs)
+        dataset = self.streaming_dataset_class(*args, **kwargs)
```
Comment (snarayan21) [inline llmfoundry/data/finetuning/tasks.py:1037]: iiuc, Streaming will throw an exception here so the code below won't even run

**Review by snarayan21 — COMMENTED** [review]
unsure if these changes actually do what we want

File: llmfoundry/data/finetuning/tasks.py (hunk 3 above)
Comment (jjanezhang) [reply llmfoundry/data/finetuning/tasks.py:1041]: We would tell them in the UI that they need to read docs and make sure they specify enough data. See [here for exact message.](https://github.com/databricks-mosaic/mcloud/blob/0d7a5066a7f30511fd4d31c2e8c1ed57fe72605a/mapi/src/graphql/runEvent/runEventCustomExceptions.ts#L174) Would this solve the issue?


The exception messages we write here are just for raw foundry users.

File: llmfoundry/command_utils/data_prep/convert_text_to_mds.py (hunk 2 above)
Comment (irenedea) [reply llmfoundry/command_utils/data_prep/convert_text_to_mds.py:481]: yup it did

File: llmfoundry/data/finetuning/tasks.py (hunk 4 above)
Comment (irenedea) [reply llmfoundry/data/finetuning/tasks.py:1037]: The error trace that @jjanezhang sent showed it erroring during data loading

Comment (irenedea) [reply llmfoundry/data/finetuning/tasks.py:1037]: Maybe num_canonical_nodes < num_samples isn't actually the right check?

```
  File "/usr/lib/python3/dist-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)  # type: ignore[possibly-undefined]
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/torch/utils/data/_utils/fetch.py", line 32, in fetch
    data.append(next(self.dataset_iter))
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/dataset.py", line 1501, in __iter__
    sample_ids = self._get_work(epoch, sample_in_epoch)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/dataset.py", line 1036, in _get_work
    epoch_sample_ids = generate_work(self.batching_method, self, p_world, epoch,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/batching/__init__.py", line 45, in generate_work
    return get(dataset, world, epoch, sample_in_epoch)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/batching/random.py", line 67, in generate_work_random_batching
    shuffle = get_shuffle(dataset.shuffle_algo, shuffle_units, dataset.num_canonical_nodes,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/shuffle/__init__.py", line 50, in get_shuffle
    return get(shard_sizes, num_canonical_nodes, seed, epoch, block_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/streaming/base/shuffle/py1e.py", line 76, in get_shuffle_py1e
    raise ValueError(f'The number of samples assigned to a canonical node is 0. This ' +
[2024-09-10 08:40:44.473000+00:00][48225]: ValueError: The number of samples assigned to a canonical node is 0. This very likely indicates that the number of samples in this stream is less than the number of canonical nodes, which is 4. Please check your index.json file and ensure that your dataset has been written out correctly. If this was intended, reduce num_canonical_nodes.

```

File: llmfoundry/command_utils/data_prep/convert_text_to_mds.py [hunk 5]
```diff
@@ -478,7 +478,9 @@ def convert_text_to_mds(
...
+                reason='No shards were created when convering text to MDS.',
```
Comment (dakinggg) [inline llmfoundry/command_utils/data_prep/convert_text_to_mds.py:482]:
```suggestion
                reason='No shards were created when converting text to MDS.',
```

**Review by dakinggg — APPROVED** [review]
lets add a unit test for raising this error?

File: llmfoundry/data/finetuning/tasks.py (hunk 3 above)
Comment (irenedea) [reply llmfoundry/data/finetuning/tasks.py:1049]: Yeah I think we would just ask them to add more data for now. Ideally we should calculate data minimums across all models to give a good solid token count that shouldn't fail.

File: llmfoundry/data/finetuning/tasks.py (hunk 4 above)
Comment (irenedea) [reply llmfoundry/data/finetuning/tasks.py:1037]: Checked and streaming does not throw an error. I also realized streaming doesn't initialize the number of canonical nodes (if it's None) until dataloading starts, so I added a check to make sure the number of samples at least the number of physical nodes.

**Pull request merged**